Audio Tinyalsa : Fixed readi API issue

kishore-sn opened the pull request:
Fixed issue of returning error from readi if requested data size is less than apb buffer size

shivgarg commented on framework/src/tinyalsa/tinyalsa.c:
```diff
@@ -553,30 +547,77 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
+		if (nbytes < pcm->nextSize) {
+			/* User wants less data than we have in buffer.
+			   Copy the requested amount of data to user buffer and return */
+			memcpy(data, apb->samp + pcm->nextOffset, nbytes);
```
Here , aren't we always wasting pcm->nextSize - nbytes of recorded data? The user may want that in a succeeding record request. I feel it is not good to discard that data. We should record less, if less amount of data is needed.

shivgarg commented on framework/src/tinyalsa/tinyalsa.c:
```diff
@@ -553,30 +547,77 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
+				return oops(pcm, errno, "failed to enque buffer after read");
+			}
+
+			return pcm_bytes_to_frames(pcm, pcm->nextSize);
```
Preserve the value of pcm->nextSize

shivgarg commented on framework/src/tinyalsa/tinyalsa.c:
```diff
@@ -553,30 +547,77 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
+				if (ioctl(pcm->fd, AUDIOIOC_ENQUEUEBUFFER, (unsigned long)&bufdesc) < 0) {
+					return oops(pcm, errno, "failed to enque buffer after read");
+				}
+				return pcm_bytes_to_frames(pcm, apb->nbytes);
```
Preserve the value of apb->nbytes.

kishore-sn commented:
Made changes suggested by Garg

Taejun-Kwon commented on framework/src/tinyalsa/tinyalsa.c:
```diff
@@ -532,14 +532,8 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
 	nbytes = pcm_frames_to_bytes(pcm, frame_count);
 
 #ifdef CONFIG_AUDIO_DRIVER_SPECIFIC_BUFFERS
```
add below in above or negative tc will fail
		if (pcm == NULL || data == NULL || frame_count == 0) {
			return -EINVAL;
		}

sunghan-chang commented on framework/src/tinyalsa/tinyalsa.c:
```diff
@@ -739,6 +779,10 @@ int pcm_close(struct pcm *pcm)
+
+	pcm->nextSize = 0;
+        pcm->nextOffset = 0;
+        pcm->nextBuf = NULL;
```
Violations of coding rule at here and above line.

kishore-sn replied:
Fixed

Taejun-Kwon commented on framework/src/tinyalsa/tinyalsa.c:
```diff
@@ -739,6 +779,10 @@ int pcm_close(struct pcm *pcm)
 	}
+
+	pcm->nextSize = 0;
+        pcm->nextOffset = 0;
```
Please align code here.

sunghan-chang replied:
align is not enough, you should use tab as an indentation. (you used spaces)

kishore-sn replied:
Fixed

Taejun-Kwon commented on framework/src/tinyalsa/tinyalsa.c:
```diff
@@ -553,29 +547,75 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
+		if (nbytes < pcm->nextSize) {
+			/* User wants less data than we have in buffer.
+			   Copy the requested amount of data to user buffer and return */
+			memcpy(data, apb->samp + pcm->nextOffset, nbytes);
```
We don't have to memcpy here.

kishore-sn replied:
We need to copy data to user even when user request smaller data .. and it is available in pending buffer...

Taejun-Kwon commented on framework/src/tinyalsa/tinyalsa.c:
```diff
@@ -553,29 +547,75 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
+		} else {
+			/* User wants more data than we have in buffer.
+			   Copy available data to user buffer, enqueue buffer and return */
+			memcpy(data, apb->samp + pcm->nextOffset, pcm->nextSize);
```
here too.

kishore-sn replied:
We need to copy data to user even when user request smaller data .. and it is available in pending buffer...

Taejun-Kwon commented on framework/src/tinyalsa/tinyalsa.c:
```diff
@@ -553,29 +547,75 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
+			   Copy the requested amount of data to user buffer and return */
+			memcpy(data, apb->samp + pcm->nextOffset, nbytes);
+			pcm->nextSize -= nbytes;
+			pcm->nextOffset += nbytes;
```
Assume that configuration is 1channel & 8bit format.
and user just call readi once with very small frame, 
then request enqueue buffer? let me hear your answer.

kishore-sn replied:
If the data frame is very small and readi is called only once..

kishore-sn commented:
Fixed indentation issues in code.

shivgarg commented:
Will the change work if there are more than 2 buffers. Here we are just book keeping the buffer which is one step ahead  but in case of more than one buffers they can be  many buffers which are dequeued but we will have lost the data recorded in them.

kishore-sn commented:
Yes. It will work for more than 2 buffers. 
We will not deque the next buffer until the current pending buffer is completely consumed. So, there will be no data loss.

shivgarg commented:
I have a doubt here regarding staleness of recorded buffers. If there is significant delay between call of pcm_readi then the later readi calls will return data that was recorded moments earlier and not at the current time. 
@kishore-sn Could you think about this?
I am writing something to overcome this and would share the code soon.

kishore-sn commented:
Updated the pull request with below changes:
- Changed variable names according to coding guidelines
- Use period_count variable and remove unwanted #ifdef CONFIG_AUDIO_DRIVER_SPECIFIC_BUFFERS
- Fixed start / stop / open API's to handle state of pcm variables
- Added pcm_drop api to behave similar to pcm_stop()
- Implemented pcm_drain API

shivgarg commented on framework/src/tinyalsa/tinyalsa.c:
```diff
@@ -531,49 +535,99 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
+			/* When we are in draining state, we will check for buffers already in the msg queue
+			   If there is no buffer on the queue, we return immediately */
+			clock_gettime(CLOCK_REALTIME, &st_time);
+			size = mq_timedreceive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio, &st_time);
```
Can we miss any message here due to some scheduling or latency issues?
Wouldn't it be better to maintain the number of buffers enqueued and then deque them in each pcm_readi while draining

kishore-sn replied:
We wont miss any buffers due to timing issues, because once a message is put on the mq, it will remain there. Also, we cannot keep count of enqueued buffers. We may enqueue 4 buffers and then call drain. By the time we call drain, the codec has filled only 2 buffers. When drain is called, codec is immediately stopped. So, Only 2 buffers can be dequeued. Since we wont know how many buffers will be filled by codec when we call drain(), we cannot keep count.

shivgarg replied:
What I meant here was if the pcm_readi calls are made quickly then the next message may not arrive at mq_timedreceive.
This message may not arrive due to large buffer size or due to latency/scheduling problems.  So we may miss the message and return oops.

What I thought of drain was to get audio data from all the enqueued buffers and to stop enqueuing buffers for further data and call pcm_stop are getting all the data. If that is not what drain means , then the in flight thing will not work.

kishore-sn replied:
Since this is in readi part, we would have already stopped the recording before we get here. After this point the driver will not send any more messages on the queue. This mq receive is done only to empty out any messages that are already present. 
However, after this point we dont need to enqueue the buffers again. That point is correct. But I just decided to keep the code simple. Even if we call enqueue, it does not matter since the codec has already been stopped and we have set our local variables like pcm->running and pcm->but_ptr to zero.

Taejun-Kwon commented on framework/include/tinyalsa/tinyalsa.h:
```diff
@@ -256,6 +256,24 @@ struct pcm;
+int pcm_drop(struct pcm *pcm);
+
+/**
+* @brief Stop a PCM preserving pending frames.
```
Please apply annotation type same as another APIs.

kishore-sn replied:
Sorry, I dont understand. I have copied the doxygen comments same as other api's.

Taejun-Kwon replied:
see pcm_open's comment. it's diffrent.

Taejun-Kwon commented on framework/src/tinyalsa/tinyalsa.c:
```diff
@@ -531,49 +535,99 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
-		return oops(pcm, EINTR, "Recieved unexpected msg (id = %d) while waiting for deque message from kernel", msg.msgId);
+		/* We got here because we dont have any pending buffers */
+		/* Wait for deque message from kernel */
+		if (pcm->draining) {
```
pcm_draining will be called after writei  finished,
you don't have to check it in readi..
Also, draining is not neccessary if it always called after writei.

Taejun-Kwon commented:
Applying pcm_drain and pcm_drop should be seperated from this commit.

Taejun-Kwon commented on framework/src/tinyalsa/tinyalsa.c:
```diff
@@ -1132,47 +1189,124 @@ int pcm_stop(struct pcm *pcm)
+ * @return On success, zero; on failure, a negative number.
+ * @ingroup libtinyalsa-pcm
+ */
+int pcm_drop(struct pcm *pcm)
```
drop is same as stop. please check when usage of drop & stop properly.
For resume, pcm_drop -> pcm_prepare will be called, I think

kishore-sn replied:
As we discussed, pcm_drop will perform same operation as pcm_stop

kishore-sn commented:
Created a new commit for pcm_drop and pcm_drain API's.

shivgarg commented on framework/src/tinyalsa/tinyalsa.c:
```diff
@@ -1075,25 +1136,24 @@ int pcm_start(struct pcm *pcm)
-		{
-			bufdesc.u.pBuffer = pcm->pBuffers[pcm->bufPtr];
+		while (pcm->buf_ptr < pcm->config.period_count) {
+			bufdesc.u.pBuffer = pcm->pBuffers[pcm->buf_ptr++];
```
I would advise to postpone this change . The current tc needs to be fixed in setting correct period parameters in set_config call. This change causes a crash with audio_tc. We need to figure out the way to manage period_size aned period_count .

kishore-sn commented:
As discussed with @Taejun-Kwon , I am closing this PR. 
I will resolve conflicts with the latest code and resubmit my code in a new PR.
